logics/impls/controller/mod.rs

```diff
@@ -718,27 +694,239 @@ impl<T: Storage<Data>> Internal for T {
         pool: AccountId,
         redeemer: AccountId,
         redeem_amount: Balance,
-        pool_attribute: Option<PoolAttributes>,
+        pool_attributes: Option<PoolAttributesForWithdrawValidation>,
     ) -> Result<()> {
-        let caller_pool = if let Some(_pool_attribute) = pool_attribute {
-            Some((pool, _pool_attribute))
-        } else {
-            None
-        };
-        let (_, shortfall) = self._get_hypothetical_account_liquidity(
-            redeemer,
-            Some(pool),
-            redeem_amount,
-            0,
-            caller_pool,
-        )?;
-        if !shortfall.is_zero() {
-            return Err(Error::InsufficientLiquidity)
-        }
+        if let Some(oracle) = self._oracle() {
+            let mut asset_price: u128 = 0;
+            let mut liquidation_threshold: u128 = 0;
+            let (shortfall, account_data) = {
+                // For each asset the account is in
+                let caller = Self::env().caller();
+                let mut skip_pool = caller;
+
+                let markets = self._markets();
+                let mut asset_params = Vec::<HypotheticalAccountLiquidityCalculationParam>::new();
+
+                let mut total_collateral_in_base_currency = U256::from(0);
+                let mut avg_ltv = U256::from(0);
+                let mut avg_liquidation_threshold = U256::from(0);
+                let mut total_debt_in_base_currency: U256 = U256::from(0);
+
+                if let Some(pool_attribute) = pool_attributes {
+                    // if caller is a pool, get parameters for the pool without call the pool
+                    if pool_attribute.underlying.is_none() {
+                        return Err(Error::UnderlyingIsNotSet)
+                    }
+                    if pool_attribute.pool.is_none() {
```

Comment:
I think we are checking the same thing.

Reply:
I think ok_or can be used since the unwrapped result is used in a subsequent step.
```rust
let underlying_in_attr =
    pool_attribute.underlying.ok_or(Error::UnderlyingIsNotSet)?;
```

Reply:
Done